### PR TITLE
qt: Use WalletModel* instead of the wallet name as map key

### PR DESCRIPTION
**Pull request by promag**
This a small refactor that doesn't change behavior. This is also necessary if in the future we allow renaming wallets.

### CODE REVIEW

File: src/qt/walletframe.h [hunk 1]
```diff
@@ -37,8 +37,8 @@ class WalletFrame : public QFrame
     void setClientModel(ClientModel *clientModel);
 
     bool addWallet(WalletModel *walletModel);
-    bool setCurrentWallet(const QString& name);
-    bool removeWallet(const QString &name);
+    bool setCurrentWallet(WalletModel* walletModel);
```
Comment (practicalswift) [inline src/qt/walletframe.h]: Make sure parameter names match between declaration and definition :-)

File: src/qt/walletframe.h [hunk 2]
```diff
@@ -37,8 +37,8 @@ class WalletFrame : public QFrame
...
+    bool removeWallet(WalletModel* walletModel);
```
Comment (practicalswift) [inline src/qt/walletframe.h]: Same here :-)

**Review by practicalswift — CHANGES_REQUESTED** [review]

**Comment by laanwj** [comment]
Concept ACK.

Small nit for clarity's sake: would make sense to word the PR title and commit as "use WalletModel pointer as map key" instead of what is *not* used, this tells reviewers more.

**Comment by promag** [comment]
@laanwj sure, actually the commit was already along that.

@practicalswift done.

**Comment by hebasto** [comment]
Concept ACK.

**Comment by jonasschnelli** [comment]
I'm not entirely convince that holding raw pointers in a list is an improvement.

**Comment by laanwj** [comment]
> @jonasschnelli what are your concerns?

I think the concern, versus using another kind of id, is that when using a pointer an inconsistency between the different data structures will cause crashes / use after free.

**Review by ryanofsky — APPROVED** [review]
utACK 492156b4b908ae1aa99b947a2efb314295d70749. This change looks like an improvement to me. Both pointers and strings should be unique but it seems easier to screw up string formatting and comparison than pointer equality checking.

In theory if there were use-after-free bugs we didn't know about, string comparisons could allow the code to limp along further without crashing, but I don't see where this could happen here because the pointers seem to be used for map lookups without getting dereferenced. Maybe you could replace `WalletModel*` with `intptr_t` to officially prevent dereferencing, though I think the `WalletModel*` is more obvious and easy to understand.

**Comment by laanwj** [comment]
So is there no other suitable unique identifier except the pointer?

I ask this because a remote RPC client would have to do with only the name too, right?

**Comment by laanwj** [comment]
> This is also necessary if in the future we allow renaming wallets.

I guess there's something to be said, then, for not ever changing that identifier while the wallet is loaded to avoid confusing remote clients, even if, say, the on-disk name or location changes.

**Comment by promag** [comment]
@laanwj my point is if there is a rename the WalletModel is the same. I'm not saying anything regarding RPC :)

**Comment by laanwj** [comment]
No, you don't say anything about RPC, but that the identifier should *probably* be constant while a wallet is loaded. Or does RPC use a different identifier for wallets than the GUI does?

**Comment by promag** [comment]
The public unique identifier is the wallet name but here `WalletModel*` is used as the internal identifier in the GUI.

Maybe I shouldn't bring up "wallet rename".. I just think using the `WalletModel*` around is more convenient than looking up model by wallet name.

**Comment by promag** [comment]
Rebased and also use `WalletModel*` in console window.

**Comment by laanwj** [comment]
travis fail:
```
qt/libbitcoinqt.a(qt_libbitcoinqt_a-rpcconsole.o): In function `RPCConsole::on_lineEdit_returnPressed()':
/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/src/qt/rpcconsole.cpp:928: undefined reference to `WalletModel::getWalletName() const'
qt/libbitcoinqt.a(qt_libbitcoinqt_a-rpcconsole.o): In function `RPCConsole::on_lineEdit_returnPressed()':
/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/src/qt/rpcconsole.cpp:928: undefined reference to `WalletModel::getWalletName() const'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by promag** [comment]
Thanks @laanwj, fixed now.

**Review by ryanofsky — APPROVED** [review]
utACK 91b0c5b09616bb8c5ef8107efb432ce4c7c9a383. Changes since last review: rebase, adding new rpc console commit, and adding copyright year updates to earlier commits.

Not a big deal, but I think rpc console commit 91b0c5b09616bb8c5ef8107efb432ce4c7c9a383 goes a little too far in passing WalletModel to parse/execute functions which really only need a wallet name to pass along to the RPC engine, and not a full WalletModel instance. It might make this code slightly more difficult to separate and test.

**Comment by promag** [comment]
@ryanofsky I had like that but RPC console already has `ENABLE_WALLET` macro to conditionally build the wallet endpoint. I think that a future refactor is to build the endpoint outside and inject it instead of passing the wallet.

**Pull request merged**